lib/charms/grafana_agent/v0/cos_agent.py

```diff
@@ -249,7 +264,206 @@ class _MetricsEndpointDict(TypedDict):
 SnapEndpoint = namedtuple("SnapEndpoint", "owner, name")
 
 
-class CosAgentProviderUnitData(pydantic.BaseModel):
+# Note: MutableMapping is imported from the typing module and not collections.abc
```

Comment:
Is this a copy of the tracing lib?
If we introduce an inter-lib dependency, could it be limited only to the grafana-agent side of things?

Reply:
It is not a copy of tracing lib as we extend the cos-agent interface with new tracing-related endpoints. In the first iteration of this PR, I tried to keep the method names similar / same as in the tracing library to reduce the effort needed to switch between directly sending traces to tempo to relaying them with gagent. However after the changes from the review it diverges a bit further so although the flow will be similar, it's not the same.

It also isn't an inter-lib dependency in terms of fetching another library. Both of them are standalone and as a charm wanting to send traces from a machine model, you won't need to import `tracing` library anymore.